File: sycl/src/bude.cpp
#include <cmath>
#include <memory>
#include <vector>
#include <chrono>
#include <iostream>
#include <fstream>
#include <functional>
#include <algorithm>
#include <CL/sycl.hpp>

#include "bude.h"

typedef std::chrono::high_resolution_clock::time_point TimePoint;

struct Params {

	size_t natlig;
	size_t natpro;
	size_t ntypes;
	size_t nposes;

	std::vector<Atom> protein;
	std::vector<Atom> ligand;
	std::vector<FFParams> forcefield;
	std::array<std::vector<float>, 6> poses;

	size_t iterations;

	// XXX bring this back once all SYCL implementations implement 2020 spec
//	size_t posesPerWI;
	size_t wgSize;
	std::string deckDir;

	clsycl::device device;

	friend std::ostream &operator<<(std::ostream &os, const Params &params) {
		os <<
		   "natlig:      " << params.natlig << "\n" <<
		   "natpro:      " << params.natpro << "\n" <<
		   "ntypes:      " << params.ntypes << "\n" <<
		   "nposes:      " << params.nposes << "\n" <<
		   "iterations:  " << params.iterations << "\n" <<
		   "posesPerWI:  " << NUM_TD_PER_THREAD << "\n" <<
		   "wgSize:      " << params.wgSize << "\n" <<
		   "SYCL device: " << params.device.get_info<clsycl::info::device::name>();
		return os;
	}
};

void fasten_main(
		clsycl::handler &h,
//		size_t posesPerWI,
		size_t wgSize,
		size_t ntypes, size_t nposes,
		size_t natlig, size_t natpro,
		clsycl::accessor<Atom, 1, R, Global> protein_molecule,
		clsycl::accessor<Atom, 1, R, Global> ligand_molecule,
		clsycl::accessor<float, 1, R, Global> transforms_0,
		clsycl::accessor<float, 1, R, Global> transforms_1,
		clsycl::accessor<float, 1, R, Global> transforms_2,
		clsycl::accessor<float, 1, R, Global> transforms_3,
		clsycl::accessor<float, 1, R, Global> transforms_4,
		clsycl::accessor<float, 1, R, Global> transforms_5,
		clsycl::accessor<FFParams, 1, R, Global> forcefield,
		clsycl::accessor<float, 1, RW, Global> etotals);


double elapsedMillis( const TimePoint &start, const TimePoint &end){
	auto elapsedNs = static_cast<double>(
			std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
	return elapsedNs * 1e-6;
}

void printTimings(const Params &params, double millis) {

	// Average time per iteration
	double ms = (millis / params.iterations);
	double runtime = ms * 1e-3;

	// Compute FLOP/s
	double ops_per_wg = NUM_TD_PER_THREAD * 27 +
			params.natlig * (2 +
					NUM_TD_PER_THREAD * 18 +
					params.natpro * (10 + NUM_TD_PER_THREAD * 30)
			) + NUM_TD_PER_THREAD;
	double total_ops = ops_per_wg * ((double) params.nposes / NUM_TD_PER_THREAD);
	double flops = total_ops / runtime;
	double gflops = flops / 1e9;

	double total_finsts = 25.0 * params.natpro * params.natlig * params.nposes;
	double finsts = total_finsts / runtime;
	double gfinsts = finsts / 1e9;

	double interactions = (double) params.nposes * (double) params.natlig * (double) params.natpro;
	double interactions_per_sec = interactions / runtime;

	// Print stats
	std::cout.precision(3);
	std::cout << std::fixed;
	std::cout << "- Kernel time:    " << (millis) << " ms\n";
	std::cout << "- Average time:   " << ms << " ms\n";
	std::cout << "- Interactions/s: " << (interactions_per_sec / 1e9) << " billion\n";
	std::cout << "- GFLOP/s:        " << gflops << "\n";
	std::cout << "- GFInst/s:       " << gfinsts << "\n";
}

template<typename T>
std::vector<T> readNStruct(const std::string &path) {
	std::fstream s(path, std::ios::binary | std::ios::in);
	if (!s.good()) {
		throw std::invalid_argument("Bad file: " + path);
	}
	s.ignore(std::numeric_limits<std::streamsize>::max());
	auto len = s.gcount();
	s.clear();
	s.seekg(0, std::ios::beg);
	std::vector<T> xs(len / sizeof(T));
	s.read(reinterpret_cast<char *>(xs.data()), len);
	s.close();
	return xs;
}

std::string deviceName(clsycl::info::device_type type) {
	//@formatter:off
	switch (type){
		case clsycl::info::device_type::cpu: return "cpu";
		case clsycl::info::device_type::gpu: return "gpu";
		case clsycl::info::device_type::accelerator: return "accelerator";
		case clsycl::info::device_type::custom: return "custom";
		case clsycl::info::device_type::automatic: return "automatic";
		case clsycl::info::device_type::host: return "host";
		case clsycl::info::device_type::all: return "all";
		default: return "(unknown: " + std::to_string(static_cast<unsigned int >(type))+ ")";
	}
	//@formatter:on
}

void printSimple(const clsycl::device &device, size_t index) {
	std::cout << std::setw(3) << index << ". "
	          << device.get_info<clsycl::info::device::name>()
	          << "(" << deviceName(device.get_info<clsycl::info::device::device_type>()) << ")"
	          << std::endl;
}

Params loadParameters(const std::vector<std::string> &args) {

	Params params = {};

	// Defaults
	params.iterations = DEFAULT_ITERS;
	params.nposes = DEFAULT_NPOSES;
	params.wgSize = DEFAULT_WGSIZE;
	params.deckDir = DATA_DIR;
//	params.posesPerWI = DEFAULT_PPWI;

	const auto readParam = [&args](size_t &current,
	                               const std::string &arg,
	                               const std::initializer_list<std::string> &matches,
	                               const std::function<void(std::string)> &handle) {
		if (matches.size() == 0) return false;
		if (std::find(matches.begin(), matches.end(), arg) != matches.end()) {
			if (current + 1 < args.size()) {
				current++;
				handle(args[current]);
			} else {
				std::cerr << "[";
				for (const auto &m : matches) std::cerr << m;
				std::cerr << "] specified but no value was given" << std::endl;
				std::exit(EXIT_FAILURE);
			}
			return true;
		}
		return false;
	};

	const auto bindInt = [](const std::string &param, size_t &dest, const std::string &name) {
		try {
			auto parsed = std::stol(param);
			if (parsed < 0) {
				std::cerr << "positive integer required for <" << name << ">: `" << parsed << "`" << std::endl;
				std::exit(EXIT_FAILURE);
			}
			dest = parsed;
		} catch (...) {
			std::cerr << "malformed value, integer required for <" << name << ">: `" << param << "`" << std::endl;
			std::exit(EXIT_FAILURE);
		}
	};

	const auto &devices = clsycl::device::get_devices();
	if (devices.empty()) {
		std::cerr << "No SYCL devices available!" << std::endl;
		std::exit(EXIT_FAILURE);
	} else {
		std::cout << "Available SYCL devices:" << std::endl;

		for (size_t j = 0; j < devices.size(); ++j) printSimple(devices[j], j);
	}

	params.device = devices[0];

	for (size_t i = 0; i < args.size(); ++i) {
		using namespace std::placeholders;
		const auto arg = args[i];
		if (readParam(i, arg, {"--iterations", "-i"}, std::bind(bindInt, _1, std::ref(params.iterations), "iterations"))) continue;
		if (readParam(i, arg, {"--numposes", "-n"}, std::bind(bindInt, _1, std::ref(params.nposes), "numposes"))) continue;
//		if (readParam(i, arg, {"--posesperwi", "-p"}, std::bind(bindInt, _1, std::ref(params.posesPerWI), "posesperwi"))) continue;
		if (readParam(i, arg, {"--wgsize", "-w"}, std::bind(bindInt, _1, std::ref(params.wgSize), "wgsize"))) continue;
		if (readParam(i, arg, {"--device", "-d"}, [&](const std::string &param) {

			auto devices = clsycl::device::get_devices();

			try { params.device = devices.at(std::stoul(param)); }
			catch (const std::exception &e) {
				std::cout << "Unable to parse/select device index `" << param << "`:" << e.what() << std::endl;
				std::cout << "Attempting to match device with substring  `" << param << "`" << std::endl;

				auto matching = std::find_if(devices.begin(), devices.end(), [param](const clsycl::device &device) {
					return device.get_info<clsycl::info::device::name>().find(param) != std::string::npos;
				});
				if (matching != devices.end()) {
					params.device = *matching;
					std::cout << "Using first device matching substring `" << param << "`" << std::endl;
				}
				else if(devices.size() == 1)
					std::cerr << "No matching device but there's only one device, will be using that anyway"<< std::endl;
				else {
					std::cerr << "No matching devices" << std::endl;
					std::exit(EXIT_FAILURE);
				}
			}
		})) { continue; }
		if (readParam(i, arg, {"--deck"}, [&](const std::string &param) { params.deckDir = param; })) continue;


		if (arg == "--list" || arg == "-l") {
			for (size_t j = 0; j < devices.size(); ++j) printSimple(devices[j], j);
			std::exit(EXIT_SUCCESS);
		}

		if (arg == "--help" || arg == "-h") {
			std::cout << "\n";
			std::cout << "Usage: ./bude [OPTIONS]\n\n"
			          << "Options:\n"
			          << "  -h  --help               Print this message\n"
			          << "  -i  --iterations I       Repeat kernel I times (default: " << DEFAULT_ITERS << ")\n"
			          << "  -n  --numposes   N       Compute energies for N poses (default: " << DEFAULT_NPOSES << ")\n"
//			          << "  -p  --poserperwi PPWI    Compute PPWI poses per work-item (default: " << DEFAULT_PPWI << ")\n"
					  << "  -w  --wgsize     WGSIZE  Run with work-group size WGSIZE using nd_range, set to 0 for plain range (default: " << DEFAULT_WGSIZE << ")\n"
			          << "  -d  --device     INDEX   Select device at INDEX from output of --list, "
			                                         "performs a substring match of device names if INDEX is not an integer (default: first device of the list)\n"
			          << "      --deck       DECK    Use the DECK directory as input deck (default: " << DATA_DIR << ")"
			          << "  -l  --list               List available devices"
			          << std::endl;
			std::exit(EXIT_SUCCESS);
		}

		std::cout << "Unrecognized argument '" << arg << "' (try '--help')" << std::endl;
		std::exit(EXIT_FAILURE);
	}

	params.ligand = readNStruct<Atom>(params.deckDir + FILE_LIGAND);
	params.natlig = params.ligand.size();

	params.protein = readNStruct<Atom>(params.deckDir + FILE_PROTEIN);
	params.natpro = params.protein.size();

	params.forcefield = readNStruct<FFParams>(params.deckDir + FILE_FORCEFIELD);
	params.ntypes = params.forcefield.size();

	auto poses = readNStruct<float>(params.deckDir + FILE_POSES);
	if (poses.size() / 6 != params.nposes) {
		throw std::invalid_argument("Bad poses: " + std::to_string(poses.size()));
	}

	for (size_t i = 0; i < 6; ++i) {
		params.poses[i].resize(params.nposes);
		std::copy(
				std::next(poses.cbegin(), i * params.nposes),
				std::next(poses.cbegin(), i * params.nposes + params.nposes),
				params.poses[i].begin());

	}

	return params;
}

template<typename T>
static clsycl::buffer<T> mkBuffer(clsycl::queue &queue, const std::vector< T> &xs ) {
	clsycl::buffer<T> buffer(xs.size());
	queue.submit([&](clsycl::handler &h) {
		h.copy(xs.data(), buffer.template get_access<RW>(h));
	});
	return buffer;
}

std::vector<float> runKernel(Params params) {

	std::vector<float> energies(params.nposes);
	auto contextStart = std::chrono::high_resolution_clock::now();
	clsycl::queue queue(params.device);
	auto contextEnd = std::chrono::high_resolution_clock::now();

	auto xferAllocStart = std::chrono::high_resolution_clock::now();

	auto protein = mkBuffer(queue, params.protein);
	auto ligand = mkBuffer(queue, params.ligand);
	auto transforms_0 = mkBuffer(queue, params.poses[0]);
	auto transforms_1 = mkBuffer(queue, params.poses[1]);
	auto transforms_2 = mkBuffer(queue, params.poses[2]);
	auto transforms_3 = mkBuffer(queue, params.poses[3]);
	auto transforms_4 = mkBuffer(queue, params.poses[4]);
	auto transforms_5 = mkBuffer(queue, params.poses[5]);
	auto forcefield = mkBuffer(queue, params.forcefield);

	clsycl::buffer<float> results(energies.size());
	queue.submit([&](clsycl::handler &h) {
		h.fill(results.get_access<W>(h), 0.f);
	});

	queue.wait();
	auto xferAllocEnd = std::chrono::high_resolution_clock::now();

	const auto runKernel = [&]() {
		queue.submit([&](clsycl::handler &h) {
			fasten_main(h,
					params.wgSize,
					params.ntypes, params.nposes,
					params.natlig, params.natpro,
					protein.get_access<R>(h),
					ligand.get_access<R>(h),
					transforms_0.get_access<R>(h),
					transforms_1.get_access<R>(h),
					transforms_2.get_access<R>(h),
					transforms_3.get_access<R>(h),
					transforms_4.get_access<R>(h),
					transforms_5.get_access<R>(h),
					forcefield.get_access<R>(h),
					results.get_access<RW>(h)
			);
		});
	};

	auto warmupStart = std::chrono::high_resolution_clock::now();
	runKernel(); //warm up
	queue.wait();
	auto warmupEnd = std::chrono::high_resolution_clock::now();

	auto kernelStart = std::chrono::high_resolution_clock::now();
	for (size_t i = 0; i < params.iterations; ++i) {
		runKernel();
	}
	queue.wait();
	auto kernelEnd = std::chrono::high_resolution_clock::now();

	clsycl::buffer<float> buffer(energies.data(), energies.size());
	queue.submit([&](clsycl::handler &h) { h.copy(results.get_access<RW>(h), buffer.get_access<RW>(h)); });
	queue.wait();

	std::cout
			<< "Context time:    " << elapsedMillis(contextStart, contextEnd) << " ms\n"
			<< "Xfer+Alloc time: " << elapsedMillis(xferAllocStart, xferAllocEnd) << " ms\n"
			<< "Warmup time:     " << elapsedMillis(warmupStart, warmupEnd) << " ms\n" << std::endl;

	printTimings(params, elapsedMillis(kernelStart, kernelEnd));
	return energies;
}

int main(int argc, char *argv[]) {

	auto args = std::vector<std::string>(argv + 1, argv + argc);
	auto params = loadParameters(args);

//	std::cout << "Parameters:\n" << params << std::endl;
	std::cout << "Device    : " << params.device.get_info<clsycl::info::device::name>() << std::endl;
	std::cout << "\tType    : " << deviceName(params.device.get_info<clsycl::info::device::device_type>()) << std::endl;
	std::cout << "\tProfile : " << params.device.get_info<clsycl::info::device::profile>() << std::endl;
	std::cout << "\tVersion : " << params.device.get_info<clsycl::info::device::version>() << std::endl;
	std::cout << "\tVendor  : " << params.device.get_info<clsycl::info::device::vendor>() << std::endl;
	std::cout << "\tDriver  : " << params.device.get_info<clsycl::info::device::driver_version>() << std::endl;
	std::cout << "Poses     : " << params.nposes << std::endl;
	std::cout << "Iterations: " << params.iterations << std::endl;
	std::cout << "Ligands   : " << params.natlig << std::endl;
	std::cout << "Proteins  : " << params.natpro << std::endl;
	std::cout << "Deck      : " << params.deckDir << std::endl;
	std::cout << "WG        : " << params.wgSize << std::boolalpha << " (use nd_range:" << (params.wgSize > 0) << ")" << std::endl;
	auto energies = runKernel(params);

	//XXX Keep the output format consistent with the C impl. so no fancy streams here
	FILE *output = fopen("energies.out", "w+");
	// Print some energies
	printf("\nEnergies\n");
	for (size_t i = 0; i < params.nposes; i++) {
		fprintf(output, "%7.2f\n", energies[i]);
		if (i < 16)
			printf("%7.2f\n", energies[i]);
	}

	// Validate energies
	std::ifstream refEnergies(params.deckDir + FILE_REF_ENERGIES);
	size_t nRefPoses = params.nposes;
	if (params.nposes > REF_NPOSES) {
		std::cout << "Only validating the first " << REF_NPOSES << " poses.\n";
		nRefPoses = REF_NPOSES;
	}

	std::string line;
	float maxdiff = 0.0f;
	for (size_t i = 0; i < nRefPoses; i++) {
		if (!std::getline(refEnergies, line)) {
			throw std::logic_error("ran out of ref energies lines to verify");
		}
		float e = std::stof(line);
		if (std::fabs(e) < 1.f && std::fabs(energies[i]) < 1.f) continue;

		float diff = std::fabs(e - energies[i]) / e;
		if (diff > maxdiff) maxdiff = diff;
	}
	std::cout << "Largest difference was " <<
	          std::setprecision(3) << (100 * maxdiff)
	          << "%.\n\n"; // Expect numbers to be accurate to 2 decimal places
	refEnergies.close();

	fclose(output);
}

File: src/julia/miniBUDE.jl/src/BUDE.jl
using ArgParse
using Parameters
using Printf
using Base: Float64, Float32, Int, Int32
import Base.read

const DefaultInterations = 8
const DefaultNPoses = 65536
const RefNPoses = 65536
const DefaultWGSize = 64
const DefaultPPWI = 4

const Zero = 0.0f0
const Quarter = 0.25f0
const Half = 0.5f0
const One = 1.0f0
const Two = 2.0f0
const Four = 4.0f0
const Cnstnt = 45.0f0

const HbtypeF = 70
const HbtypeE = 69
const Hardness = 38.0f0
const Npnpdist = 5.5f0
const Nppdist = 1.0f0

const Float32Max = typemax(Float32)

struct Vec3f32
  x::Float32
  y::Float32
  z::Float32
end

struct Vec4f32
  x::Float32
  y::Float32
  z::Float32
  w::Float32
end

struct Atom
  x::Float32
  y::Float32
  z::Float32
  type::Int32
end

function read(io::IO, ::Type{Atom})
  Atom(read(io, Float32), read(io, Float32), read(io, Float32), read(io, Int32))
end

struct FFParams
  hbtype::Int32
  radius::Float32
  hphb::Float32
  elsc::Float32
end

function read(io::IO, ::Type{FFParams})
  FFParams(read(io, Int32), read(io, Float32), read(io, Float32), read(io, Float32))
end

@with_kw mutable struct Params
  device::String = "1"
  list::Bool = false
  numposes::UInt = DefaultNPoses
  iterations::UInt = DefaultInterations
  wgsize::UInt = DefaultWGSize
  ppwi::UInt = DefaultPPWI
  deck::String = "../data/bm1"
end

struct Deck
  protein::AbstractArray{Atom}
  ligand::AbstractArray{Atom}
  forcefield::AbstractArray{FFParams}
  poses::AbstractArray{Float32,2}
end

const DeviceWithRepr = Tuple{Any,String,Any}

function parse_options(given::Params)
  s = ArgParseSettings()
  @add_arg_table s begin
    "--list", "-l"
    help = "List available devices"
    action = :store_true
    "--device", "-d"
    help = "Select device at DEVICE or the device name substring if DEVICE is not an integer, NOTE: Julia is 1-indexed"
    arg_type = String
    default = given.device
    "--numposes", "-n"
    help = "Compute energies for N poses"
    arg_type = Int
    default = convert(Int, given.numposes)
    "--iterations", "-i"
    help = "Repeat kernel I times"
    arg_type = Int
    default = convert(Int, given.iterations)
    "--wgsize", "-w"
    help = "Run with work-group size WGSIZE"
    arg_type = Int
    default = convert(Int, given.wgsize)
    "--ppwi", "-p"
    help = "Compute PPWI poses per work-item"
    arg_type = Int
    default = convert(Int, given.ppwi)
    "--deck"
    help = "Use the DECK directory as input deck"
    arg_type = String
    default = given.deck
  end
  args = parse_args(s)
  # surely there's a better way than doing this:
  for (arg, val) in args
    setproperty!(given, Symbol(arg), val)
  end
end

function read_structs(path::String, ::Type{T})::Vector{T} where {T}
  io = open(path, "r")
  size = filesize(io)
  [read(io, T) for _ = 1:size÷sizeof(T)]
end

function print_timings(params::Params, deck::Deck, elapsedSeconds::Float64, wgsize::UInt)

  # Average time per iteration
  averageIterationSeconds = elapsedSeconds / params.iterations

  natlig = length(deck.ligand)
  natpro = length(deck.protein)

  # Compute FLOP/s
  ops_per_wg = (wgsize * 27 + natlig * (2 + wgsize * 18 + natpro * (10 + wgsize * 30)) + wgsize)

  total_ops = ops_per_wg * (params.numposes / wgsize)
  flops = total_ops / averageIterationSeconds
  gflops = flops / 1.0e9

  total_finsts = 25.0 * natpro * natlig * params.numposes
  finsts = total_finsts / averageIterationSeconds
  gfinsts = finsts / 1e9

  interactions = params.numposes * natlig * natpro
  interactions_per_sec = interactions / averageIterationSeconds

  # Print stats
  @printf "- Kernel time:    %.03f ms\n" (elapsedSeconds * 1000.0)
  @printf "- Average time:   %.03f ms\n" (averageIterationSeconds * 1000.0)
  @printf "- Interactions/s: %.03f billion\n" interactions_per_sec / 1e9
  @printf "- GFLOP/s:        %.03f\n" gflops
  @printf "- GFInst/s:       %.03f\n" gfinsts
end

function main()

  params::Params = Params()
  parse_options(params)


  ds = devices()

  if params.list
    for (i, (_, name, type)) in enumerate(ds)
      println("[$i] ($type) $name")
    end
    exit(0)
  end

  deviceIndex = try
    index = parse(Int, params.device)
    if index < 1 || index > length(ds)
      error("Device $(index) out of range (1..$(length(ds))), NOTE: Julia is 1-indexed")
    end
    index
  catch
    index = findfirst(x -> occursin(params.device, "($(x[3])) $(x[2])"), ds)
    if index === nothing
      error("No device match the substring `$(params.device)`, see --list for available devices")
    end
    index
  end


  poses = permutedims( # reshape is column order so we flip it back again
    reshape(
      read_structs("$(params.deck)/poses.in", Float32),  # read poses as one long array
      (params.numposes, 6), # reshape it to 6 slices of numposes sized rows
    ),
  )

  deck = Deck(
    read_structs("$(params.deck)/protein.in", Atom),
    read_structs("$(params.deck)/ligand.in", Atom),
    read_structs("$(params.deck)/forcefield.in", FFParams),
    poses,
  )

  println("Poses     : ", size(deck.poses)[2])
  println("Iterations: ", params.iterations)
  println("Ligands   : ", length(deck.ligand))
  println("Protein   : ", length(deck.protein))
  println("Forcefield: ", length(deck.forcefield))
  println("Deck      : ", params.deck)
  println("WGsize    : ", params.wgsize)
  println("PPWI      : ", params.ppwi)

  set_zero_subnormals(true)
  GC.enable(false)

  (energies, rumtimeSeconds, ppwi) = run(params, deck, ds[deviceIndex])

  GC.enable(true)
  set_zero_subnormals(false)

  print_timings(params, deck, rumtimeSeconds, ppwi)

  output = open("energies.out", "w+")
  println("\nEnergies:")
  for i = 1:size(deck.poses)[2]
    @printf(output, "%7.2f\n", energies[i])
    if (i < 16)
      @printf("%7.2f\n", energies[i])
    end
  end

  ref_energies = open("$(params.deck)/ref_energies.out", "r")
  n_ref_poses = size(deck.poses)[2]
  if n_ref_poses > RefNPoses
    println("Only validating the first $(RefNPoses) poses")
    n_ref_poses = RefNPoses
  end

  maxdiff = 0.0
  for i = 1:n_ref_poses
    line = readline(ref_energies)
    if line == ""
      error("ran out of ref energies lines to verify")
    end
    e = parse(Float32, line)
    if (abs(e) < 1.0 && abs(energies[i]) < 1.0)
      continue
    end
    diff = abs(e - energies[i]) / e
    if (diff > 0.01)
      println("![$(i)] $(e) $(energies[i]) ~ $(diff)")
      maxdiff = diff
    end
  end

  @printf "Largest difference was %.03f%%.\n\n" 100 * maxdiff # Expect numbers to be accurate to 2 decimal places
  close(ref_energies)

end
